File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

 /*
perform inorder traversal to get sorted arr of nodes
return kth element
1. s=[]
n=0
curr=root
2. while (stack.length > 0 || curr) {
    while (curr) {
        stack.push(curr)
        curr = curr.left
    }
    curr = stack.pop()
    n++
    if (n == k) return curr.val
    curr = curr.right
}
 */

/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    var curr = root
    var n = 0
    var stack = []

    while (curr || stack.length > 0) {
        while (curr) {
            stack.push(curr)
            curr = curr.left
        }
        curr = stack.pop()
        n++

        if (n == k) {
            return curr.val
        }

        curr = curr.right
    }
};

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        s = []
        curr = root
        n = 0

        while len(s) > 0 or curr:
            while curr:
                s.append(curr)
                curr = curr.left

            node = s.pop()
            n += 1

            if n == k:
                return node.val
    
            curr = node.right
